src/containers/app/listItemsProvider.js

```diff
@@ -0,0 +1,44 @@
+import React, { useReducer } from "react";
+import { ToDoContext } from "./context";
+
+const reducer = ( state, action ) => {
+  switch (action.type) {
+    case 'addToList': 
+      return addToList(state, action.payload);
+    case 'deleteFromList':
+      return deleteFromList(state, action.payload);
+    case 'editFromList':
+      return editFromList(state, action.payload);
+    default:
+      throw new Error('Unexpected action'); 
+  }
+}
+
+const addToList = (state, inputVal) => {
+  return ([...state, { id: Date.now(), text: inputVal }]);
+};
+
+const deleteFromList = (state, id) => {
+  return (state.filter((task) => task.id !== id));
+};
+
+const editFromList = (state, { id, newValue }) => {
+   return (
+    state.map((task) => {
+      return task.id === id ? { id, text: newValue } : task;
+    })
+  );
+};
```

Comment:
It's better to define functions before using them.
In such cases it will work fine because reducer function will be called after those functions will be defined, but in other cases it can broke something.

you can define functions after using them when you use "function" operator.

say('ok')
function(text) { alert(text) } // Works fine

say('ok')
const say = (text) => { alert(text) } // Uncaught ReferenceError: Cannot access 'say' before initialization